File: Ordination/Python/SelectionSort.py
#coding: utf-8
from Sort import Sort
class SelectionSort(Sort):
	pass
	'''
		Generic Method
		algorithm.sort(array) -> return array complete ordered
		algorithm.sort(array,leftIndex,rightIndex) -> return array partially ordered
	'''
	def sort(self,array,leftIndex = 0 , rightIndex = None):
		if(rightIndex == None):
			rightIndex = len(array) -1
		if self.__validateParams(array,leftIndex,rightIndex):
			#self.__simpleSelectionSort(array,leftIndex,rightIndex)
			#self.__bidirectionalSelectionSort(array,leftIndex,rightIndex)
			#self.__recursiveSimpleSelectionSort(array,leftIndex,rightIndex)
			self.__recursiveBidirectionalSelectionSort(array,leftIndex,rightIndex)
	
	'''
		Simple iteractive selection sort implementation
	'''
	def __simpleSelectionSort(self,array,leftIndex,rightIndex):
		while(leftIndex < rightIndex):
			minIndex = leftIndex;
			for i in range(leftIndex+1,rightIndex+1):
				if(array[minIndex]> array[i]):
					minIndex = i
			array[minIndex],array[leftIndex] = array[leftIndex] ,array[minIndex]
			leftIndex = leftIndex + 1
	
	'''
		Bidirectional iteractive selection sort implementation
	'''
	def __bidirectionalSelectionSort(self,array,leftIndex,rightIndex):
		while(leftIndex < rightIndex):
			minIndex = leftIndex
			for i in range(leftIndex + 1 ,rightIndex +1):
				if(array[minIndex] > array[i]):
					minIndex = i		
			array[minIndex], array[leftIndex] = array[leftIndex],array[minIndex]
			leftIndex = leftIndex +1
			
			maxIndex = rightIndex
			for i in range(rightIndex-1 , leftIndex-1,-1):
				if(array[maxIndex] < array[i]):
					maxIndex = i
			array[maxIndex] , array[rightIndex] = array[rightIndex],array[maxIndex]
			rightIndex = rightIndex -1

	'''
		Recursive simple  selection sort implementation
	'''
	def __recursiveSimpleSelectionSort(self,array,leftIndex,rightIndex):
			if(leftIndex < rightIndex):
				self.__recursiveGetMin(array,leftIndex,rightIndex,leftIndex,rightIndex)
				self.__recursiveSimpleSelectionSort(array,leftIndex + 1,rightIndex)

	'''
		Recursive bidirectional selection sort implementation
	'''
	def __recursiveBidirectionalSelectionSort(self,array,leftIndex,rightIndex):
			if(leftIndex < rightIndex):
				self.__recursiveGetMax(array,leftIndex,rightIndex,rightIndex,rightIndex)
				self.__recursiveGetMin(array,leftIndex,rightIndex,leftIndex,leftIndex)
				self.__recursiveBidirectionalSelectionSort(array,leftIndex + 1,rightIndex-1)

	'''
		This method get max element in array and switch with rightIndex position
	'''
	def __recursiveGetMax(self,array,leftIndex,rightIndex,maxValue,index):
		if(index >= leftIndex):
			if(array[maxValue] < array[index]):
				maxValue = index
			self.__recursiveGetMax(array,leftIndex,rightIndex,maxValue,index-1)
		else:
			array[maxValue],array[rightIndex] = array[rightIndex] , array[maxValue]

	'''
		This method get min element in array and switch with leftIndex position
	'''
	def __recursiveGetMin(self,array,leftIndex,rightIndex,minValue,index):
		if(index <= rightIndex):
			if(array[minValue] > array[index]):
				minValue = index
			self.__recursiveGetMin(array,leftIndex,rightIndex,minValue,index +1)
		else:
			array[minValue] ,array[leftIndex] = array[leftIndex] , array[minValue]

	'''
		This method verify if parameters are valid if some parameter is not valid the method returns false	
	'''
	def __validateParams(self,array,leftIndex,rightIndex):
		if(array == None or leftIndex < 0 or rightIndex > len(array) -1 or leftIndex >= rightIndex):
			return False
		else:
			return True;

File: Ordination/C++/SelectionSort.cpp
#include <iostream>


template<typename T> class SelectionSort{
	public:
		void sort(T array[],int leftIndex,int rightIndex,int size){
			if(validateParams(array,leftIndex,rightIndex,size)){
				//this->simpleSelectionSort(array,leftIndex,rightIndex);
				//this->bidirectionalSelectionSort(array,leftIndex,rightIndex);
				this->recursiveSimpleSelectionSort(array,leftIndex,rightIndex);
				//this->recursiveBidirectionalSelectionSort(array,leftIndex,rightIndex);
				this->showArray(array,size);
			}else{
				std::cout << "Entrou na area errada!!" << std::endl;
			}
		}	
		void sort(T array[],int size){
			this->sort(array,0,size - 1 ,size);
		}
	private:

		void simpleSelectionSort(T array[],int leftIndex,int rightIndex){
			while(leftIndex < rightIndex){
				int minIndex = leftIndex;
				for(int j = leftIndex + 1; j <= rightIndex;j++){
					if(array[minIndex] > array[j]){
						minIndex = j;
					}
				}
				swap(array,leftIndex++,minIndex);
			}
		}
		
		void bidirectionalSelectionSort(T array[],int leftIndex,int rightIndex){
			while(leftIndex < rightIndex){
				int minIndex = leftIndex;
				for(int j = leftIndex + 1;j <= rightIndex;j++){
					if(array[minIndex] > array[j]){
						minIndex = j;
					}
				}
				
				swap(array,leftIndex++,minIndex);
				
				int maxIndex = rightIndex;
				for(int i = rightIndex; i >= leftIndex;i--){
					if(array[maxIndex] < array[i]){
						maxIndex = i;
					}
				}
				swap(array,rightIndex--,maxIndex);
			}
		}
		
		void recursiveSimpleSelectionSort(T array[],int leftIndex,int rightIndex){
			if(leftIndex < rightIndex){
				this->recursiveGetMin(array,leftIndex,rightIndex,leftIndex,leftIndex);
				this->recursiveSimpleSelectionSort(array,leftIndex +1,rightIndex);
			}	
		}
		void recursiveBidirectionalSelectionSort(T array[],int leftIndex,int rightIndex){
			if(leftIndex < rightIndex){
				this->recursiveGetMin(array,leftIndex,rightIndex,leftIndex,leftIndex);
				this->recursiveGetMax(array,leftIndex,rightIndex,rightIndex,rightIndex);
				this->recursiveBidirectionalSelectionSort(array,leftIndex + 1,rightIndex- 1);
			}
		}
		void recursiveGetMin(T array[],int leftIndex,int rightIndex,int min,int index){
			if(index <= rightIndex){
				if(array[min] >array[index]){
					min = index;
				}
				this->recursiveGetMin(array,leftIndex,rightIndex,min,index + 1);
			}else{
				swap(array,leftIndex,min);
			}
		}
	
		void recursiveGetMax(T array[],int leftIndex,int rightIndex,int max, int index){
			if(index >= leftIndex){
				if(array[max] < array[index]){
					max = index;
				}
				this->recursiveGetMax(array,leftIndex,rightIndex,max,index -1);
			}else{
				swap(array,rightIndex,max);
			}
		}
		bool validateParams(T array[],int leftIndex,int rightIndex,int size){
			if(array == NULL || leftIndex < 0 || rightIndex > size - 1 || leftIndex >= rightIndex){
				return false;
			}else{
				return true;
			}
		}
		
		void swap(T array[],int i ,int j){
			T aux = array[i];
			array[i] = array[j];
			array[j] = aux;
		}	
		
		void showArray(T array[],int size){
			std::cout << "Array:  [";
			for(int i = 0 ; i < size; i++){
				if(i != size - 1){
					std::cout << array[i] << ",";
				}else{
					std::cout << array[i] << "]";	
				}
			}
			std::cout << std::endl;
		}
};

